### DISCUSSION TITLE
Async support

### DISCUSSION THREAD

**Discussion by shomodj** [pawamoy/duty#28] [Ideas]
Interesting tool, I like that it is similar to invoke, I just wish it supported async. Do you think this is complex to add? @pawamoy

**Comment by pawamoy** [comment]
Interesting idea @shomodj! Could you show a use-case where async support would be useful?

**Reply by shomodj** [reply]
for example I wanted to create a task that will ssh/scp on a remote machine and I wanted to use asyncssh, I know its a bit silly but when you write a whole project using async I also want build/task tools to be able to use that code

**Reply by shomodj** [reply]
BTW none of them have it, invoke and click don't work with async as far as I know

**Reply by pawamoy** [reply]
I see, thanks! So basically you'd like to define async functions that work as duties, right?

```python
@duty
async def my_task(ctx):
    ...
```

Have you tried running your coroutines directly with `asyncio.run`?

```python
@duty
def my_task(ctx):
    ctx.run(asyncio.run, args=[your_coroutine])
```

**Reply by pawamoy** [reply]
More concrete example:

```python

import asyncio

# let say you have an async func
async def async_task() -> None:
    print("Sleeping...")
    await asyncio.sleep(1)
    print("done!")

@duty
def async_duty(ctx: Context) -> None:
    # run it like that:
    ctx.run(asyncio.run, args=[async_task()], title="Running an async task")
```

**Reply by shomodj** [reply]
yes, exactly, this will support asyncio without any shenanigans

**Reply by pawamoy** [reply]
I suppose you mean "yes, `async def my_duty` so that I don't have to `asyncio.run`"? OK! I'm sure we can do the `asyncio.run` part ourselves when running the duty :slightly_smiling_face: Doesn't sound like a complicated change. Can you open a feature request?

**Reply by shomodj** [reply]
sorry, yes

```
@duty
async def task(ctx: Context) -> None:
    await async.sleep(1)
```
